source/llassetgen/source/Image.cpp

```diff
@@ -184,40 +195,63 @@ namespace llassetgen {
 			abort();
 		}
 
-		if (setjmp(png_jmpbuf(png))){
+		if (setjmp(png_jmpbuf(png))) {
 			std::cerr << "pnglib caused a longjump due to an error" << std::endl;
 			std::cerr << "could not write file " << filepath;
 			abort();
 		}
 
 		png_set_write_fn(png, (png_voidp)&out_file, write_data, flush_data);
 
-		png_set_IHDR(	png, 
-						info, 
-						width, 
-						height,
-						bit_depth, 
-						PNG_COLOR_TYPE_GRAY, 
-						PNG_INTERLACE_NONE,
-						PNG_COMPRESSION_TYPE_BASE, 
-						PNG_FILTER_TYPE_BASE);
+		png_set_IHDR(png,
+			info,
+			width,
+			height,
+			(bit_depth <= 16) ? bit_depth : 16,
+			PNG_COLOR_TYPE_GRAY,
+			PNG_INTERLACE_NONE,
+			PNG_COMPRESSION_TYPE_BASE,
+			PNG_FILTER_TYPE_BASE);
 
 		png_write_info(png, info);
 
 		if (bit_depth > 8) {
 			png_set_swap(png);
 		}
-	
-		for (int y = 0; y < height; y++) {
-			png_write_row(png, reinterpret_cast<png_bytep>(&data.get()[y * stride]));
+
+		if (bit_depth >= 24) {
+			std::unique_ptr<uint16_t[]> row(new uint16_t[width]);
+			// possible 32 float or 32 or 24 bit int data
+			// scale down to 16 bit int grayscale
+			
+			if (max > pow(2, bit_depth)) {
+				max = pow(2, bit_depth);
```

Comment:
pow(2, bit_depth) is using floating point arithmetics
Use bit shifting instead: 1 << bit_depth